battle-engine.js

```diff
@@ -697,18 +696,24 @@ class BattlePokemon {
 
 		return true;
 	}
-	formeChange(template, dontRecalculateStats) {
+	formeChange(template, source) {
 		template = this.battle.getTemplate(template);
 
 		if (!template.abilities) return false;
+
+		template = this.battle.singleEvent('ModifyTemplate', this.battle.getFormat(), null, this, source, null, template);
+
+		if (!template) return false;
+
 		this.template = template;
 
 		this.types = template.types;
-		this.addedType = '';
+		this.addedType = template.addedType || '';
 		this.knownType = true;
 
-		if (!dontRecalculateStats) {
+		if (!source) {
 			let stats = this.battle.spreadModify(this.template.baseStats, this.set);
+			if (!this.baseStats) this.baseStats = stats;
```

Comment:
Slightly confusing I guess given line 719 but I need this in order to set the base hp.